### ISSUE TITLE
grad, vjp transforms have weird interaction with torch.no_grad

### ISSUE THREAD

**Issue by zou3519** [facebookresearch/functorch#79]
Should torch.no_grad affect the grad/vjp transforms? Why or why not?

**Comment by zou3519** [comment]
One really annoying thing about torch.no_grad is that it is not traceable. JAX has a stop_gradient primitive that operates on Tensors so it does become traceable: https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.stop_gradient.html

**Comment by zou3519** [comment]
Proposal:
- torch.no_grad does affect grad/vjp transforms. Any computation that happens within torch.no_grad is invisible to vjp/grad
- If a user calls grad/vjp inside of torch.no_grad, we raise a warning that explains that their gradients will be 0. (Or maybe this should be an error?)
- For tracing... either we introduce something like `stop_gradient(Tensor) -> Tensor` or figure out how to "trace" `torch.no_grad`. This sounds a bit like factory function tracing and could potentially be done with a mode-based dispatch key

**Comment by zou3519** [comment]
This is a duplicate of #13, closing this
